src/autolabel/generator.py

```diff
@@ -0,0 +1,126 @@
+        )
+
+    def _generate_column_description(self, col: str, examples: List[Dict]) -> str:
+        prompt = f"Please provide a concise one sentence description for the column '{col}'. Here are some examples of items in this column: {examples}\nDescription:"
+        result = self.llm.label([prompt])
+        if result.errors[0] is not None:
+            raise result.errors[0]
+        description = result.generations[0][0].text.strip()
+        return description
+
+    def generate(
+        self,
+        num_rows: int,
+        guidelines: str = DEFAULT_GUIDELINES,
+        output_guidelines: str = DEFAULT_OUTPUT_GUIDELINES,
+        provide_examples: bool = True,
+        column_descriptions: Dict[str, str] = {},
+    ) -> pd.DataFrame:
+        """
+        This method generates a synthetic dataset based on the provided guidelines and column descriptions.
+
+        Args:
+            num_rows (int): The number of rows to generate in the dataset.
+            guidelines (str, optional): The guidelines for generating the dataset. Defaults to DEFAULT_GUIDELINES.
+            provide_examples (bool, optional): Whether to provide examples for generating the dataset. Defaults to True.
+            column_descriptions (Dict[str, str], optional): Descriptions for each column in the dataset. Defaults to an empty dictionary.
+
+        Returns:
+            pd.DataFrame: The generated synthetic dataset.
+        """
+        example_set = None
+        if provide_examples and self.config.few_shot_example_set() is not None:
+            if isinstance(self.config.few_shot_example_set(), List):
+                example_set = self.config.few_shot_example_set()
+            else:
+                example_set = pd.read_csv(
+                    self.config.few_shot_example_set(), sep=self.config.delimiter()
+                )
+                example_set = example_set.sample(
+                    min(len(example_set), self.config.few_shot_num_examples())
+                )
+                example_set = example_set.to_dict(orient="records")
+        else:
+            example_set = None
+        examples = self._format_examples(example_set)
+
+        output_guidelines = output_guidelines.format(
+            num_rows=num_rows, columns=self.columns
+        )
+
+        prompt = PROMPT_TEMPLATE.format(
+            guidelines=guidelines,
+            examples=examples,
+            output_guidelines=output_guidelines,
+            column_descriptions="\n".join(
+                [
+                    f"- {col}: {column_descriptions[col] if col in column_descriptions else self._generate_column_description(col, example_set)}"
+                    for col in self.columns
+                ]
+            ),
+        )
+
+        logger.info(f"Prompt: {prompt}")
+
+        result = self.llm.label([prompt])
+        if result.errors[0] is not None:
+            raise result.errors[0]
+
+        response = result.generations[0][0].text.strip()
```

Comment:
If it ends with this markdown, will it also not start with this in many cases?

Reply:
I use the start of the markdown as a sort of starting point for the LLM. The prompt passed to the LLM ends with "\`\`\`csv" to encourage the LLM start with the csv texts instead of anything else. From testing, the LLM is unlikely to repeat "```csv" at the beginning of its response.